Libraries/LibURL/Pattern/PatternParser.cpp

```diff
@@ -0,0 +1,405 @@
+    // NOTE: This was a "{foo}" grouping. We add this to the pending fixed value so that it will be combined with
+    //       any previous or subsequent text.
+    if (!name_token && !regexp_or_wildcard_token && modifier == Part::Modifier::None) {
+        // 1. Append prefix to the end of parser’s pending fixed value.
+        m_pending_fixed_value.append(prefix);
+
+        // 2. Return.
+        return {};
+    }
+
+    // 4. Run maybe add a part from the pending fixed value given parser.
+    TRY(maybe_add_a_part_from_the_pending_fixed_value());
+
+    // 5. If name token is null and regexp or wildcard token is null:
+    // NOTE: This was a "{foo}?" grouping. The modifier means we cannot combine it with other text. Therefore we
+    //       add it as a part immediately.
+    if (!name_token && !regexp_or_wildcard_token) {
+        // 1. Assert: suffix is the empty string.
+        VERIFY(suffix.is_empty());
+
+        // 2. If prefix is the empty string, then return.
+        if (prefix.is_empty())
+            return {};
+
+        // 3. Let encoded value be the result of running parser’s encoding callback given prefix.
+        auto encoded_value = TRY(m_encoding_callback(prefix));
+
+        // 4. Let part be a new part whose type is "fixed-text", value is encoded value, and modifier is modifier.
+        // 5. Append part to parser’s part list.
+        m_part_list.append({ Part::Type::FixedText, move(encoded_value), modifier });
+
+        // 6. Return.
+        return {};
+    }
+
+    // 6. Let regexp value be the empty string.
+    // NOTE: Next, we convert the regexp or wildcard token into a regular expression.
+    String regexp_value;
+
+    // 7. If regexp or wildcard token is null, then set regexp value to parser’s segment wildcard regexp.
+    if (!regexp_or_wildcard_token) {
+        regexp_value = m_segment_wildcard_regexp;
+    }
+    // 8. Otherwise if regexp or wildcard token’s type is "asterisk", then set regexp value to the full wildcard regexp value.
+    else if (regexp_or_wildcard_token->type == Token::Type::Asterisk) {
+        regexp_value = ".*"_string;
+    }
+    // 9. Otherwise set regexp value to regexp or wildcard token’s value.
+    else {
+        regexp_value = regexp_or_wildcard_token->value;
+    }
+
+    // 10. Let type be "regexp".
+    // NOTE: Next, we convert regexp value into a part type. We make sure to go to a regular expression first so
+    //       that an equivalent "regexp" token will be treated the same as a "name" or "asterisk" token.
+    auto type = Part::Type::Regexp;
+
+    // 11. If regexp value is parser’s segment wildcard regexp:
+    if (regexp_value == m_segment_wildcard_regexp) {
+        // 1. Set type to "segment-wildcard".
+        type = Part::Type::SegmentWildcard;
+
+        // 2. Set regexp value to the empty string.
+        regexp_value = String {};
+    }
+    // 12. Otherwise if regexp value is the full wildcard regexp value:
+    else if (regexp_value == ".*"sv) {
```

Comment:
Given that we repeat this constant quite a lot, maybe put it in a static var somewhere? I keep finding myself checking that we have the right string value :sweat_smile: